src/validation.cpp

```diff
@@ -609,17 +609,53 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         size_t nLimitAncestorSize = gArgs.GetArg("-limitancestorsize", DEFAULT_ANCESTOR_SIZE_LIMIT)*1000;
         size_t nLimitDescendants = gArgs.GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
         size_t nLimitDescendantSize = gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
+
+        if (setConflicts.size() == 1) {
+            // In general, when we receive an RBF transaction with mempool conflicts, we want to know whether we
+            // would meet the chain limits after the conflicts have been removed. However, there isn't a practical
+            // way to do this short of calculating the ancestor and descendant sets with an overlay cache of
+            // changed mempool entries. Due to both implementation and runtime complexity concerns, this isn't
+            // very realistic either, thus we we try to ensure a limited set of transactions are RBF'able despite
+            // mempool conflicts here. Importantly, we want to ensure that transactions which were accepted using
+            // the below carve-out are able to be RBF'ed, without impacting the security the carve-out provides
+            // for off-chain contract systems (see link in the comment below).
+            //
+            // Specifically, the subset of RBF transactions which we allow despite chain limits are those which
+            // conflict directly with exactly one other transaction (but may evict children of said transaction),
+            // and which are not adding any new mempool dependencies. Note that the "no new mempool dependencies"
+            // check is accomplished later, so we don't bother doing anything about it here, but if BIP 125 is
+            // amended, we may need to move that check to here instead of removing it wholesale.
+            //
+            // Such transactions are clearly not merging any existing packages, so we are only concerned with
+            // ensuring that (a) no package is growing past the package size (not count) limits and (b) we are
+            // not allowing something to effectively use the (below) carve-out spot when it shouldn't be allowed
+            // to.
+            //
+            // To check these we first check if we meet the RBF criteria, above, and increment the descendant
+            // limits by the single direct conflict (as these are recalculated by assuming the new transaction
+            // being added is a new dependant, with no removals, of each parents' existing dependant set).
+            // The ancestor count limits are unmodified (as the ancestor limits should be the same for both our
+            // new transaction and any conflicts).
+            assert(setIterConflicting.size() == 1);
+            CTxMemPool::txiter conflict = *setIterConflicting.begin();
+
+            nLimitDescendants += 1;
+            nLimitDescendantSize += conflict->GetTxSize();
```

Comment:
note the comment above would have to change too https://github.com/bitcoin/bitcoin/pull/16421/files#diff-24efdb00bfbe56b140fb006b562cc70bR641

Reply:
Done! good idea.

Reply:
Can't we actually cope with merging packages this way too though? If you've got a tx that has parents A, B, C; and that conflicts with tx's X (parent A) and Y (parent B), then beforehand you had:

```
..., A, X, [children X]
..., B, Y, [children Y]
```

(maybe some tx's were descendants of both X and Y, but it's worse if there weren't any like that) and afterwards you have:

```
..., A, tx
..., B, tx
```

You don't have C in the mempool because that would fail the "replacement-adds-unconfirmed" test later.

So you know tx's ancestor checks pass, because they're actually worked out; you know A,B's ancestor checks pass because they don't change, tx's descendant check is trivial, and you know A,B and all their parent's descendant checks passed, because they did so when X and Y were there -- as far as sizes go, if they were all at their limit, then the max size for tx is the minimum of the descendant sizes of each tx that was replaced.

So I think you could replace the `setConflicts.size() == 1` test with:

```
if (!setConflicts.empty()) {
    auto conflict = setIterConflicting.begin();
    assert(conflict != setIterConflicting.end());
    uint64_t bump = (*conflict)->GetSizeWithDescendants();
    while(++conflict != setIterConflicting.end()) {
        bump = std::min(bump, (*conflict)->GetSizeWithDescendants());
    }
    nLimitDescendants += 1;
    nLimitDescendantSize += bump;
}
```

Maybe I'm missing something though?

Reply:
I agree there's lots we can do in the future to improve things (not actually 100% sure on this one, these things are complicated...) lets try to land this as-is first, though, cause I do really want it for 0.19 still.